### PR TITLE
Create coding-style.md

### PR DESCRIPTION
**Pull request by oddcoder**
This is just a draft, possibly incomplete one, I tried to keep it as short as possible. We can always add extra guidelines as the need arise, please  let me know what you think. Please let me know what you think.

### CODE REVIEW

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Is there any way we can standardize around just using `*` for text decoration? We anyways have to `**` if we want to bold, so can we just say that `*` is for italics instead of using two different text decoration styles in markdown (here and elsewhere in this doc)?

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Can we reflow the markdown text so that it's at max 80 characters? It's a bit hard to read if you don't have wordwrap (here and the rest of this document).

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Text flow here is a bit strange, maybe:

```
Try to avoid: multiple empty lines, comment boxes, and excessive comments in function bodies.
```

should be one sentence on its own.

I generally agree with "you need to separately comment parts of it, it should probably be refactored"; but sometimes that's not totally practical in certain complex cases. I don't think we should be this restrictive about things - we should say to prefer this, but not always demand it.

There's going to be cases where a function is kind of complicated and splitting it up actually hurts readability instead of helps. I'm okay with suggesting split-up, but not sure I'm on-board with requiring it to avoid comments.

File: coding-style.md
Comment (ghost) [inline coding-style.md]: I think we should split traits and macros into separate sections. They have their own complex things to talk about. My 2 cents would be to say the following:

### Traits

* Traits on public types expose an API feature that a client could depend on, so they should be considered carefully - don't implement a trait just for convenience.
* Manual trait implementations should appear after impl blocks for a type.
* Try to keep trait impls ordered by the name trait that it is implementing (where possible).
* If there are multiple traits of the same name, try to ordered further by the generic typename present (if possible).

### Macros

* Derive macros should share the name of the trait that they're implementing.
* Macros should try to contain their logic as much as possible, so that no other types in the same module can reach into the guts of privately defined functions on the types.
* If a macro is closely related to another type (for example it implements a bunch of functions or traits on a type and the macro is used for code deduplication) the invocation of the macro should appear very close to the type definition that it is invoked on.
* Try to avoid `macro_rules` unless you need it - the rust type system is often good enough for sharing implementation details with generic impls. (Unfortunately sometimes you'll need `macro_rules`, but the point is you don't always need it to share implementation, it's not the only way of doing this).

File: coding-style.md
Comment (ghost) [inline coding-style.md]: This might be pedantic, but for section headings can we avoid having a period on the end so that we treat them more like a title, so that they're always "Title Casing" and usually drop the ending punctuation (here and elsewhere in this doc)?

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Commits or Pull Requests?

File: coding-style.md
Comment (ghost) [inline coding-style.md]: I don't understand what you mean by this.

File: coding-style.md
Comment (ghost) [inline coding-style.md]: I notice some of the code organization changes that I mentioned in https://github.com/orgs/tpm-rs/discussions/18 aren't present here. Should we add them, or should I send a PR on top of this PR to add them later?

File: coding-style.md
Comment (ghost) [inline coding-style.md:12]: I don't really see what we're borrowing from this.

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Nit: Can you capitalize "Previous" since it's the start of a new sentence?

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Well, when we PR we squash and rebase, so I don't know how important this is to say since the PR itself must build, run tests, satisfy formatting checks and clippy checks, etc.

Maybe more clearly we should focus on the fact that one PR always should result in one commit by the time its pushed.

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Maybe "Code Structure" is a better heading name? "Patch" makes it sound like we're talking about git stuff here.

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Maybe we should expand on this a bit more, `//!` and `///` does something very different from `//`. Also, does this need restated? There's already rules about commenting in https://github.com/rust-lang/style-team - maybe we should just let be the guidance (for example they allow comments within certain cases to be `/**/`, such as when marking up a parameter, like `/*param=*/foo`).

File: coding-style.md
Comment (ghost) [inline coding-style.md]: We should also talk about dev branches, what should the naming format of them be, or should we prefer that folks do development in their forks. It's been a bit of a mix right now, I don't care either way, but if we allow dev branches in tpm-rs then I'd at least like to dictate a naming style (my preference is `dev/<user-name>/<branch-name>`).

Thoughts?

File: coding-style.md
Comment (ghost) [inline coding-style.md]: Actually, I also don't see what we're borrowing from this, either.

**Review by ghost — COMMENTED** [review]
Thanks for the doc!

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: right, I somehow totally missed it, will add the ones we all agreed on and feel free to suggest more as you see fit.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: So I personally use forks, also to avoid accidentally pushing stuff I don't want to push, but I wouldn't mind someone having their own dev branch. What I would say is we shouldn't encourage it ? but if someone explicitly ask for doing that (or start doing it), we can revisit the topic, or we can just explicitly ask people to do dev work on their own fork. what do you think?

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: > Maybe we should expand on this a bit more, //! and /// does something very different from //.

I really hoped people would know the difference, if they didn't, I believe rustc/clippy will catch some of the mistakes and we can catch the rest.

> There's already rules about commenting in https://github.com/rust-lang/style-team , maybe we should just let be the guidance

You are right, but I would say this particular rule is on the [list of rust style guidelines](https://github.com/rust-lang/rust/tree/HEAD/src/doc/style-guide/src#comments). The reason I mentioned it is because, linting tools wouln't catch that, and I would prefer not to have to argue with someone to replace their huge block comments with `//` when I can just refer them to our guidelines.

> maybe we should just let be the guidance (for example they allow comments within certain cases to be /**/, such as when marking up a parameter, like /*param=*/foo).

I wouldn't advocate for this particular use case since rls can already show those markups and update them on the fly, but I understand the general point you are trying to make. Please note that this is not strict prohibition of block comments. It is just preference that every guideline suggests but not many people follow. So in that particular case you suggested, line comments wouldn't have worked, so it would be fine to use block comments as they work.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md:12]: So It will be hard to point one rule to one place, but I might have copied stuff from all links above, sometimes word for word. I prefer not to have anyone claiming we copied their guides without proper credits

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: That is taken almost word for word, from the kernel contribution guide. I can actually make good case for the need for extreme simplicity, in our particular case. It should be very simple for anyone who reads the business logic to fact check it against the specs.  That being said, simple functions does not have to be short, they can be long. Maybe I will try to reword it and give the rewording here. Or I can expand it like the kernel people did:

> Also, try to avoid putting comments inside a function body: if the function is so complex that you need to separately comment parts of it, you should probably go back to chapter 6 for a while. You can make small comments to note or warn about something particularly clever (or ugly), but try to avoid excess. Instead, put the comments at the head of the function, telling people what it does, and possibly WHY it does it.

What do you think?

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: I would try not to say "don't implement a trait just for convenience" because  I think that is the whole point of traits to offer convenience. just putting a note to myself as well. let me know what you think.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: so, we had a discussion at https://github.com/tpm-rs/tpm-rs/issues/66 about this

You can read through it, the case that preempted me to request that was when I was sending bunch of refactoring to `marshal`.  My idea was that to send many commit per PR but each of them would be treated individually, so it is convenient for the sender.  So everyone preferred not to do that, and instead send small PRs, but we agreed to do it only when necessary and hoped people won't abuse that. So that is why I said commits instead of PRs. It is the only case so far where I think sending multiple commits might make sense.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: relates to commit vs pull request, I added a section to allow multiple commits but prefer single one. If you are totally against it, please let me know, it is very possible that I am just biased towards that.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: So this describes the following situation:
Contributor submitted 10 patches together in the first patch they introduce function `foo`. and maybe in the third patch they modify the function `foo` because their initial implementation wasn't 100% suitable for the usecase.

Ideally one would want to have the first patch introduce the full complete function `foo` that wouldn't modified shortly after. I am not sure if that helps clarifies things.

fwiw, this was actually copied from rustc contributor guidelines. But I can try to reword it if you think we should.

**Comment by oddcoder** [comment]
I think the document should be ready for review now, I addressed all your comments (I think), feel free to point out anything I missed. Also, everything here is up for discussion if you want something to be added or removed, please let me know. Also please note this is not the final version the document, If you agree with the general guidelines and directions, I will give it another review to make the document style more consistent.

File: coding-style.md
Comment (ghost) [reply coding-style.md]: I agree with extreme simplicity, and generally I'm in favour of it. I have seen this taken to the extreme though, and in ways that has hurt code instead of helped. I'm fine with the current wording, and we can iterate on this if there are issues with PRs later on (though I doubt it, I think most folks will have good sense for when to split logic or not).

LGTM - thanks!

File: coding-style.md
Comment (ghost) [reply coding-style.md]: I'll also add that I don't really like the idea of reviewing each individual commit in a PR. That sounds pretty difficult actually, and if the PR is huge potentially very time-consuming. I'd rather only consider the final diff after applying all the commits and then squashing them down.

That said, I'm fine with keeping the experiment open for allowing non-squashed PRs if you find value in them, I just don't and honestly may even forget to review the different commits individually to see if they're sensible.

The way that I like to do PRs is just take a bunch of commits, and that's your working "PR", and then when we go to accept the PR it has to be squashed down.

If it's a dev-branch force pushing is fine, folks shouldn't be taking dependencies on other folks dev branches anyways. So I don't see an issue with that either, personally.

---

Anyways, we can keep the wording as "commits" here to account for us allowing non-squashed rebase+ff PRs.

File: coding-style.md
Comment (ghost) [reply coding-style.md]: Nah this is fine, I think that if we just squash and rebase it doesn't usually apply and this edge-case goes away (it kind of just becomes "the code is allowed to evolve", which is fine to say) - but it makes sense to keep around if we're going to allow non-squashed PRs.

File: coding-style.md
Comment (ghost) [reply coding-style.md:12]: :+1:

File: coding-style.md
Comment (ghost) [reply coding-style.md]: I'm pretty against it, but not enough to be forceful about it. I just think it's going to be hard to review PRs, but I could be wrong about that.

This PR is a great example - in this case it's a markdown doc so it doesn't matter that much, but I really don't want to click through all these commits and see that each of them makes sense in isolation.

I'm against non-squashed PRs but not strongly enough to push about it. We can try it for a while longer, and just see what happens. :)

File: coding-style.md
Comment (ghost) [reply coding-style.md]: > I really hoped people would know the difference, if they didn't, I believe rustc/clippy will catch some of the mistakes and we can catch the rest.

That's fair, I'm being pedantic here.

---

My comment wasn't really specifically about block comments, it was just about us repeating what something we already linked to said already. But repeating it for emphasis or clarity is fine.

As long as we discuss the cases where they do seem to provide value in the future then I'm fine with this. I think usually avoiding them is the right call, and instead preferring to break up code though.

File: coding-style.md
Comment (ghost) [reply coding-style.md]: I don't really mind working from dev branches as long as the main branch is protected. The problem I personally have with it is then we have to maintain and clean up after folks' dev branches if they move on or abandon their work (https://github.com/tpm-rs/tpm-rs/branches/all).

So probably preference for forks makes the most sense to me. But I guess we don't have to be so authoritative about it. Let's just revisit this in the future if it should become a problem.

**Comment by ghost** [comment]
This sounds good to me - let's go with this and then we can point to this when there are disagreements. I'd like to confirm though that we should remain willing to open up discussion on these items in the future if anything doesn't feel right. Maybe the process for that is to open an issue in this docs repo?

I'd also like to note that the docs repo is private, and that's not super great because I'd want to be able to point to this for public contributors. Is this repo being private intended?

**Comment by ghost** [comment]
I apparently cannot approve this since I'm not a reviewer, but this comment is my informal approval.

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: I reworked the text and Assumed it is alway 1 PR == 1 Commit almost everywhere

File: coding-style.md
Comment (oddcoder) [reply coding-style.md]: I haven't noticed until now, I personally use my fork, so I don't have one opinion or the other.

**Comment by oddcoder** [comment]
I reworded couple of things, in particular normalizing 1 PR = 1 patch + all additional modification to that 1 patch. 

I added few guidelines regarding unsafe code and feature flags, and did my best to eliminate all  spelling mistakes.

@chrisfenner when you have some free time, please give it a look and approve?
Also it would be nice  you can:
- add @trent-reed and myself to the list of reviewers so we can commit code.
- make this repository public ?

The idea is this there are lots of ongoing discussions, about various parts of tpm-rs and it might be nice to move the final design of particular module here. Of course that is easier to do when the module is just designed than later. However at last me, I wouldn't put too much effort into it.

**Review by ghost — APPROVED** [review]
Oh, sorry it looks like I can approve - my bad, I'm still a bit new to GitHub specifically; I've used git in the past but not on this platform.

Anyways, here's an approval, hopefully it does something! :)

**Comment by oddcoder** [comment]
> Oh, sorry it looks like I can approve - my bad, I'm still a bit new to GitHub specifically; I've used git in the past but not on this platform.

Nope, as it seams none of us have write access here:

image

**Review by chrisfenner — APPROVED** [review]
Approving based on @trent-reed approval while we sort out updates to MAINTAINERS

**Comment by oddcoder** [comment]
@chrisfenner sorry for having to DM you again but apparently I cannot even merge to this  repo
image

**Comment by chrisfenner** [comment]
> @chrisfenner sorry for having to DM you again but apparently I cannot even merge to this repo image

oof! Let's fix this asap. I'll merge this now to unblock.

Sorry for the delay, I was on vacation.

**Pull request merged**
